pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,937 @@
+	if svc.Spec.Type == corev1.ServiceTypeClusterIP {
+		return internalTrafficCluster
+	}
+	externalTrafficCluster := svc.Spec.ExternalTrafficPolicy == corev1.ServiceExternalTrafficPolicyTypeCluster
+	return internalTrafficCluster && externalTrafficCluster
+}
+
+func (c *Controller) addEndpointSlice(obj interface{}) {
+	eps := obj.(*discovery.EndpointSlice)
+	svc, _ := c.serviceLister.Services(eps.GetNamespace()).Get(eps.GetLabels()[discovery.LabelServiceName])
+	if svc == nil {
+		return
+	}
+	// Events for EndpointSlices dedicated to Services without a `Local` traffic policy are ignored, as the Service IPs
+	// will always be advertised.
+	if noLocalTrafficPolicy(svc) {
+		return
+	}
+	if c.hasAffectedBPByService(svc) {
+		klog.V(2).InfoS("Processing EndpointSlice ADD event", "EndpointSlice", klog.KObj(eps))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) updateEndpointSlice(_, obj interface{}) {
+	eps := obj.(*discovery.EndpointSlice)
+	svc, _ := c.serviceLister.Services(eps.GetNamespace()).Get(eps.GetLabels()[discovery.LabelServiceName])
+	if svc == nil {
+		return
+	}
+	// Events for EndpointSlices dedicated to Services without a `Local` traffic policy are ignored, as the Service IPs
+	// will always be advertised.
+	if noLocalTrafficPolicy(svc) {
+		return
+	}
+	if c.hasAffectedBPByService(svc) {
+		klog.V(2).InfoS("Processing EndpointSlice UPDATE event", "EndpointSlice", klog.KObj(eps))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) deleteEndpointSlice(obj interface{}) {
+	eps := obj.(*discovery.EndpointSlice)
+	svc, _ := c.serviceLister.Services(eps.GetNamespace()).Get(eps.GetLabels()[discovery.LabelServiceName])
+	if svc == nil {
+		return
+	}
+	// Events for EndpointSlices dedicated to Services without a `Local` traffic policy are ignored, as the Service IPs
+	// will always be advertised.
+	if noLocalTrafficPolicy(svc) {
+		return
+	}
+	if c.hasAffectedBPByService(svc) {
+		klog.V(2).InfoS("Processing EndpointSlice DELETE event", "EndpointSlice", klog.KObj(eps))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) hasAffectedBPByEgress() bool {
+	allBPs, _ := c.bgpPolicyLister.List(labels.Everything())
+	for _, bp := range allBPs {
+		if !c.matchedCurrentNode(bp) {
+			continue
+		}
+		if bp.Spec.Advertisements.Egress != nil {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *Controller) addEgress(obj interface{}) {
+	eg := obj.(*v1beta1.Egress)
+	if eg.Status.EgressNode != c.nodeName {
+		return
+	}
+	if c.hasAffectedBPByEgress() {
+		klog.V(2).InfoS("Processing Egress ADD event", "Egress", klog.KObj(eg))
+		c.queue.Add(dummyKey)
+	}
+}
+
+func (c *Controller) updateEgress(oldObj, obj interface{}) {
+	oldEg := oldObj.(*v1beta1.Egress)
+	eg := obj.(*v1beta1.Egress)
+	if oldEg.Status.EgressNode != c.nodeName && eg.Status.EgressNode != c.nodeName {
+		return
+	}
+	if oldEg.Status.EgressIP == eg.Status.EgressIP {
+		return
+	}
```

Comment:
This caused the Egress IP not to be adversited when it migrates.

Reply:
Done